### PR TITLE
[netcore] Improve support for rare enum types

### PR DESCRIPTION
**Pull request by filipnavara**
Improve support for rare enum types (IntPtr, UIntPtr, Single, Double, Char, Boolean backing types) and make `Enum.HasFlag` working.

`System.Runtime.Tests  Total: 32076, Errors: 0, Failed: 122, Skipped: 0, Time: 29.682s`

Actual number of fixed tests is quite high because the discovery was broken. Remaining `Enum` related failures:

```
    System.Tests.EnumTests.Parse<TestName_Single>(value: "vaLue2", ignoreCase: True, expected: Value1) [FAIL]
      Assert.Equal() Failure
      Expected: Value1
      Actual:   Value2
      Stack Trace:
        /Users/vsts/agent/2.149.2/work/1/s/src/System.Runtime/tests/System/EnumTests.cs(133,0): at System.Tests.EnumTests.Parse[TestName_Single](String value, Boolean ignoreCase, TestName_Single expected)
        /Users/filipnavara/Documents/mono/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs(391,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Tests.EnumTests.Parse<TestName_Double>(value: "vaLue2", ignoreCase: True, expected: Value1) [FAIL]
      Assert.Equal() Failure
      Expected: Value1
      Actual:   Value2
      Stack Trace:
        /Users/vsts/agent/2.149.2/work/1/s/src/System.Runtime/tests/System/EnumTests.cs(133,0): at System.Tests.EnumTests.Parse[TestName_Double](String value, Boolean ignoreCase, TestName_Double expected)
        /Users/filipnavara/Documents/mono/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs(391,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Tests.EnumTests.Parse_NetCoreApp11<TestName_Single>(value: "vaLue2", ignoreCase: True, expected: Value1) [FAIL]
      Assert.Equal() Failure
      Expected: Value1
      Actual:   Value2
      Stack Trace:
        /Users/vsts/agent/2.149.2/work/1/s/src/System.Runtime/tests/System/EnumTests.netcoreapp.cs(28,0): at System.Tests.EnumTests.Parse_NetCoreApp11[TestName_Single](String value, Boolean ignoreCase, TestName_Single expected)
        /Users/filipnavara/Documents/mono/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs(391,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Tests.EnumTests.Parse_NetCoreApp11<TestName_Double>(value: "vaLue2", ignoreCase: True, expected: Value1) [FAIL]
      Assert.Equal() Failure
      Expected: Value1
      Actual:   Value2
      Stack Trace:
        /Users/vsts/agent/2.149.2/work/1/s/src/System.Runtime/tests/System/EnumTests.netcoreapp.cs(28,0): at System.Tests.EnumTests.Parse_NetCoreApp11[TestName_Double](String value, Boolean ignoreCase, TestName_Double expected)
        /Users/filipnavara/Documents/mono/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs(391,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

### CODE REVIEW

**Comment by filipnavara** [comment]
Looks like the remaining tests are wrong, they basically test that CoreCLR parses the values incorrectly.

**Comment by marek-safar** [comment]
@vargaz please review

File: mcs/class/System.Private.CoreLib/System/Enum.cs [hunk 1]
```diff
@@ -41,10 +41,17 @@ public override bool Equals (object obj)
 			return DefaultEquals (this, obj);
 		}
 
-		[Intrinsic]
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		private extern bool InternalHasFlag(Enum flags);
+
 		public bool HasFlag (Enum flag)
```
Comment (vargaz) [inline mcs/class/System.Private.CoreLib/System/Enum.cs]: This is an JIT intrinsic in mono, so it should still have the [Intrinsic] attribute.

**Comment by vargaz** [comment]
Are float/double really valid enum basetypes ? The spec says it needs to be a built-in integer type.

File: mcs/class/System.Private.CoreLib/System/Enum.cs (hunk 1 above)
Comment (filipnavara) [reply mcs/class/System.Private.CoreLib/System/Enum.cs]: It wasn't treated as such and the non-netcore Mono build also uses `InternalHasFlag`.

**Comment by filipnavara** [comment]
> Are float/double really valid enum basetypes ?

Yes, but only on .NET Core. It doesn't always work correctly, but the tests expect it to be supported.

File: mcs/class/System.Private.CoreLib/System/Enum.cs (hunk 1 above)
Comment (filipnavara) [reply mcs/class/System.Private.CoreLib/System/Enum.cs]: I'll can add it back, but shouldn't it be in mscorlib too?

Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System/Enum.cs]: It doesn't matter much, we don't check for it anywhere, its just a marker.

**Pull request merged**